evennia/contrib/turnbattle.py

```diff
@@ -0,0 +1,520 @@
+        """
+        You'll probably want to replace this with your own system for recovering HP.
+        """
+
+"""
+----------------------------------------------------------------------------
+SCRIPTS START HERE
+----------------------------------------------------------------------------
+"""
+
+class TurnHandler(Script):
+    """
+    This is the script that handles the progression of combat through turns.
+    On creation (when a fight is started) it adds all combat-ready characters
+    to its roster and then sorts them into a turn order. There can only be one
+    fight going on in a single room at a time, so the script is assigned to a
+    room as its object.
+    
+    Fights persist until only one participant is left with any HP or all
+    remaining participants choose to end the combat with the 'disengage' command.
+    """
+    
+    def at_script_creation(self):
+        """
+        Called once, when the script is created.
+        """
+        self.key = "Combat Turn Handler"
+        self.interval = 1 # Once a second
+        self.persistent = True
+        self.db.fighters = []
+        # Add all fighters in the room with at least 1 HP to the combat."
+        for object in self.obj.contents:
+            if object.db.hp:
+                self.db.fighters.append(object)
+        # Initialize each fighter for combat
+        for fighter in self.db.fighters:
+            combat_cleanup(fighter) #Clean up leftover combat attributes beforehand, just in case.
+            fighter.db.Combat_ActionsLeft = 1 #Actions remaining - start of turn adds to this, turn ends when it reaches 0
+            fighter.db.Combat_TurnHandler = self #Add a reference to this script to the character
+            fighter.db.Combat_LastAction = "null" #Track last action taken in combat
+        # Roll initiative and sort the list of fighters depending on who rolls highest to determine turn order.
+        # The initiative roll is determined by the roll_init function and can be customized easily.
+        ordered_by_roll = sorted(self.db.fighters, key=roll_init, reverse=True)
+        self.db.fighters = ordered_by_roll
+        # Announce the turn order.
+        self.obj.msg_contents("Turn order is: %s " % ", ".join(obj.key for obj in self.db.fighters))
+        "Set up the current turn and turn timeout delay."
+        self.db.turn = 0
+        self.db.timer = 30 # 30 seconds
+        
+    def at_stop(self):
+        """
+        Called at script termination.
+        """
+        for fighter in self.db.fighters:
+            combat_cleanup(fighter) #Clean up the combat attributes for every fighter.
+    
+    def at_repeat(self):
+        """
+        Called once every self.interval seconds.
+        """
+        currentchar = self.db.fighters[self.db.turn] # Note the current character in the turn order.
+        self.db.timer -= 1 # Count down the timer by one second.
```

Comment:
Ticking every second is a a simple but wasteful way to handle a countdown. The only times that the script cares about are after 20s (when the warning happens) and 30s (when the next round is force-started). You could therefore get away with ticking the script only every 10 seconds instead. You use the scripts' `restart` method whenever you want to prematurely start the next round on-demand. You can simply have your action-set commands check in to the script whenever an action is performed, so it knows when the actions are used up.

I'm not necessarily requiring you to rewrite this, but doing things on-demand is the recommended way to handle timeouts in an asynchronous system like Evennia. Checking things over and over for change is generally a bad practice.